### PR TITLE
[13.0][ADD] product_supplierinfo_qty_multiplier: Add Multiplier Qty to vendor pricelist in products to auto-set correct qty in purchase order lines

### PR DESCRIPTION
**Pull request by victoralmau**
Add Multiplier Qty to vendor pricelist in products to auto-set correct qty in purchase order lines.

Please @pedrobaeza and @joao-p-marques can you review it?

@Tecnativa TT30744

### CODE REVIEW

File: product_supplierinfo_qty_multiplier/models/product_supplierinfo.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+# Copyright 2021 Tecnativa - Pedro M. Baeza
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import _, api, exceptions, fields, models
+
+
+class ProductSupplierinfo(models.Model):
+    _inherit = "product.supplierinfo"
+
+    multiplier_qty = fields.Float(string="Multiplier Qty", required=True, default=1)
```
Comment (legalsylvain) [inline product_supplierinfo_qty_multiplier/models/product_supplierinfo.py]: Hi. Setting ``required=True and default=1`` + adding ``raise Warning`` prevent users to set custom values, that is a big limitation, and make the module working only for product with a ``Unit`` UoM.
Simply removing both parts will make the module generic. 

(and so also removing ``set_default_multiplier_qty_supplierinfo``)

File: product_supplierinfo_qty_multiplier/models/purchase_order_line.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        params = {"order_id": self.order_id}
+        seller = self.product_id._select_seller(
+            partner_id=self.partner_id,
+            quantity=self.product_qty,
+            date=self.order_id.date_order and self.order_id.date_order.date(),
+            uom_id=self.product_uom,
+            params=params,
+        )
+        qty = self.product_qty
+        if seller and int(qty / seller.multiplier_qty) != qty / seller.multiplier_qty:
```
Comment (legalsylvain) [inline product_supplierinfo_qty_multiplier/models/purchase_order_line.py]: ``        if seller and seller.multiplier_qty and qty % self.multiplier_qty:``

File: product_supplierinfo_qty_multiplier/models/purchase_order_line.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+            self.product_qty = ceil(qty / seller.multiplier_qty) * seller.multiplier_qty
+            return {
+                "warning": {
+                    "title": _("Warning"),
+                    "message": _(
+                        "The selected supplier only sells this product by %s %s"
+                    )
+                    % (seller.multiplier_qty, seller.product_uom.name),
```
Comment (legalsylvain) [inline product_supplierinfo_qty_multiplier/models/purchase_order_line.py]: maybe add a text in the message to say that the quantity has been changed automatically. (and says the new quantity)

**Review by legalsylvain — CHANGES_REQUESTED** [review]
Hi @victoralmau. Thanks a lot for this module. I have developped some years ago an incubated module here that has same feature. (https://github.com/grap/grap-odoo-incubator/tree/12.0/purchase_package_qty)
glad to have the possibility to switch to a community solution.

However, a blocking point. 

otherwise, LGTM.

thanks !

File: product_supplierinfo_qty_multiplier/models/product_supplierinfo.py (hunk 1 above)
Comment (pedrobaeza) [reply product_supplierinfo_qty_multiplier/models/product_supplierinfo.py]: @victoralmau I think it's better to add a boolean field instead saying "Only multipliers". If so, the quantity indicated in the "Min. qty" field will be the multiplier.

Comment (legalsylvain) [reply product_supplierinfo_qty_multiplier/models/product_supplierinfo.py]: Hi @pedro. I see what you mean, but there are some limitation in your implementation. 
Min qty is also to mention a price associated to a min qty.
Let's see : 
- product beer (75cl)
- supplierinfo 1 : min_qty : 0, multiplier_qty : 6 ; price : 2,40€
- supplierinfo 2 : min_qty : 300, multiplier_qty : 6 ; price : 2,20€

if you add simply a boolean, you remove such valid configuration.

Comment (pedrobaeza) [reply product_supplierinfo_qty_multiplier/models/product_supplierinfo.py]: OK, I see. Then it's good to have both indeed. Let's go with the suggestions of not being required. If not filled (0), no multiplier.

**Comment by victoralmau** [comment]
Changes done.

**Review by pedrobaeza — APPROVED** [review]

**Comment by pedrobaeza** [comment]
@legalsylvain is it ok for you?

File: product_supplierinfo_qty_multiplier/README.rst [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+.. contents::
+   :local:
+
+Usage
+=====
+
+To use this module, you need to:
+
+#. Go to Purchase > Products > Products.
+#. Go to Acoustic Bloc Screens product.
```
Comment (joao-p-marques) [inline product_supplierinfo_qty_multiplier/README.rst]: It's better not to give usage instructions depending on demo data:
```suggestion
#. Select a product.
```
(If the selected product needs some special condition, specify it)

File: product_supplierinfo_qty_multiplier/README.rst [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+#. Go to 'Purchase' tab and edit 'Wood Corner' line to set 'Multiplier Qty' to 2.0.
```
Comment (joao-p-marques) [inline product_supplierinfo_qty_multiplier/README.rst]: Again, don't restrict it to one product, and warn the user that the column might be hidden.
```suggestion
#. Go to the 'Purchase' tab and edit a vendor pricelist line to set 'Multiplier Qty' to 2.0 (If you can't find that column, you can activate it by clicking the tree dots at the corner of the table and enabling it).
```

File: product_supplierinfo_qty_multiplier/README.rst [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+#. Create a new purchase order.
+#. Add product line with 'Acoustic Bloc Screens' product product.
```
Comment (joao-p-marques) [inline product_supplierinfo_qty_multiplier/README.rst]:
```suggestion
#. Add a product line with the previously edited product.
```

File: product_supplierinfo_qty_multiplier/models/purchase_order_line.py [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+
+class PurchaseOrderLine(models.Model):
+    _inherit = "purchase.order.line"
+
+    @api.onchange("product_qty")
+    def _onchange_quantity(self):
+        super()._onchange_quantity()
+        if not self.product_id:
+            return
```
Comment (joao-p-marques) [inline product_supplierinfo_qty_multiplier/models/purchase_order_line.py]:
```suggestion
        res = super()._onchange_quantity()
        if not self.product_id:
            return res
```

File: product_supplierinfo_qty_multiplier/models/purchase_order_line.py [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+        params = {"order_id": self.order_id}
+        seller = self.product_id._select_seller(
+            partner_id=self.partner_id,
+            quantity=self.product_qty,
+            date=self.order_id.date_order and self.order_id.date_order.date(),
+            uom_id=self.product_uom,
+            params=params,
+        )
+        qty = self.product_qty
+        if seller and seller.multiplier_qty and qty % seller.multiplier_qty:
```
Comment (joao-p-marques) [inline product_supplierinfo_qty_multiplier/models/purchase_order_line.py:27]: You can combine these two I think. @pedrobaeza correct me if I'm wrong...
```suggestion
        if seller.multiplier_qty and qty % seller.multiplier_qty:
```

File: product_supplierinfo_qty_multiplier/models/purchase_order_line.py [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+                        "The quantity has been automatically changed to %s %s"
+                    )
+                    % (
+                        seller.multiplier_qty,
+                        seller.product_uom.name,
+                        self.product_qty,
+                        seller.product_uom.name,
+                    ),
+                }
+            }
```
Comment (joao-p-marques) [inline product_supplierinfo_qty_multiplier/models/purchase_order_line.py]:
```suggestion
            res.update({
                "warning": {
                    "title": _("Warning"),
                    "message": _(
                        "The selected supplier only sells this product by %s %s.\n"
                        "The quantity has been automatically changed to %s %s"
                    )
                    % (
                        seller.multiplier_qty,
                        seller.product_uom.name,
                        self.product_qty,
                        seller.product_uom.name,
                    ),
                }
            })
            return res
```

File: product_supplierinfo_qty_multiplier/views/product_supplierinfo_view.xml [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+            <label for="price" position="before">
+                <label for="multiplier_qty" />
+                <div class="o_row">
+                    <field name="multiplier_qty" />
+                </div>
+            </label>
+        </field>
+    </record>
+    <record id="product_supplierinfo_tree_view" model="ir.ui.view">
+        <field name="name">product.supplierinfo.form.view</field>
```
Comment (joao-p-marques) [inline product_supplierinfo_qty_multiplier/views/product_supplierinfo_view.xml]:
```suggestion
        <field name="name">product.supplierinfo.tree.view</field>
```

File: product_supplierinfo_qty_multiplier/README.rst [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+
+Usage
+=====
+
+To use this module, you need to:
+
+#. Go to Purchase > Products > Products.
+#. Go to Acoustic Bloc Screens product.
+#. Go to 'Purchase' tab and edit 'Wood Corner' line to set 'Multiplier Qty' to 2.0.
+#. Create a new purchase order.
```
Comment (joao-p-marques) [inline product_supplierinfo_qty_multiplier/README.rst:44]:
```suggestion
#. Create a new purchase order.
#. Assign it to the partner where the multiplier quantity was previously set.
```

**Review by joao-p-marques — COMMENTED** [review]
Code and functional review. LGTM, just some minor comments.

File: product_supplierinfo_qty_multiplier/models/purchase_order_line.py (hunk 8 above)
Comment (pedrobaeza) [reply product_supplierinfo_qty_multiplier/models/purchase_order_line.py:27]: `seller` can be `False`, not an empty recordset, so it's correct as it is.

File: product_supplierinfo_qty_multiplier/README.rst [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+   :local:
+
+Usage
+=====
+
+To use this module, you need to:
+
+#. Go to Purchase > Products > Products.
+#. Select a product.
+#. Go to the 'Purchase' tab and edit a vendor pricelist line to set 'Multiplier Qty' to 2.0 (If you can't find that column, you can activate it by clicking the tree dots at the corner of the table and enabling it).
```
Comment (joao-p-marques) [inline product_supplierinfo_qty_multiplier/README.rst]: Sorry, I mislead you earlier, there is a typo here:
```suggestion
#. Go to the 'Purchase' tab and edit a vendor pricelist line to set 'Multiplier Qty' to 2.0 (If you can't find that column, you can activate it by clicking the three dots at the corner of the table and enabling it).
```

**Review by joao-p-marques — APPROVED** [review]
Thanks @victoralmau. LGTM now, just a minor typo...

**Pull request merged**